glitch/parsers/docker_parser.py

```diff
@@ -0,0 +1,294 @@
+        stages = {}
+        for i, stage_i in enumerate(stage_indexes):
+            stage_image = structure[stage_i].value.split(" ")[0]
+            stage_start = stage_i if i != 0 else 0
+            stage_end = len(structure) if i == len(stage_indexes) - 1 else stage_indexes[i + 1]
+            stages[stage_image] = DockerParser.__get_stage_structure(structure, stage_start, stage_end)
+        return stages
+
+    @staticmethod
+    def __get_stage_structure(structure: list[DFPStructure], stage_start: int, stage_end: int):
+        sub_structure = structure[stage_start:stage_end].copy()
+        DockerParser.__add_user_tag(sub_structure)
+        return sub_structure
+
+    @staticmethod
+    def __create_variable_block(element: DFPStructure) -> Variable:
+        v: Variable
+        if element.instruction == 'USER':
+            v = Variable("user", element.value, False)
+        elif element.instruction == 'ARG':
+            value = element.value.split("=")
+            arg = value[0]
+            default = value[1] if len(value) == 2 else None
+            v = Variable(arg, default if default else "ARG", True)
+        elif element.instruction == 'ENV':  # ENV Missing support for multiple values
+            split_char = "=" if "=" in element.value else " "
+            assert len(element.value.split(split_char)) == 2
+            env, value = element.value.split(split_char)
+            v = Variable(env, value, "$" in value)
+        else:  # LABEL
+            label, value = element.value.split("=")
+            v = Variable(label, value, False)
+
+        v.line = element.startline + 1
+        return v
+
+    @staticmethod
+    def __has_user_tag(structure: list[DFPStructure]) -> bool:
+        return bool(s for s in structure if s.instruction == "USER")
+
+    @staticmethod
+    def __add_user_tag(structure: list[DFPStructure]):
+        if len([s for s in structure if s.instruction == "USER"]) > 0:
+            return
+
+        index, line = -1, -1
+        for i, s in enumerate(structure):
+            if s.instruction == 'FROM':
+                index = i
+                line = s.startline
+                break
+        structure.insert(index + 1, DFPStructure("USER root", line, "USER", line, "root"))
+
+    @staticmethod
+    def __merge_download_commands(unit_block: UnitBlock):
+        download_units = [a for a in unit_block.atomic_units if DockerParser.__is_download(a)]
+        for d in download_units:
+            file = DockerParser.__get_atomic_file_name(d)
+            d.add_attribute(Attribute("url", d.name, False))
+            d.name = ""
+            f_manipulations = [au for au in unit_block.atomic_units if DockerParser.__get_atomic_file_name(au) == file]
+
+            for au in f_manipulations:
+                if au.type in ["gpg", "checksum", "md5sum"]:
+                    d.add_attribute(Attribute(au.type, "true", False))
+                    unit_block.atomic_units.remove(au)
```

Comment:
Not merging the downloads, if there is a download I search for those 3 operations (gpg, checksum, md5sum) that affect the downloaded file and if I find any of them I add them as an attribute to the download, this is to add the integrity check to an attribute in the download since it is not a option in the wget command

Reply:
I think I am not fully understanding. Could you provide an example? It seems to me that you are mixing the detection of the smells with the parsing and that may not be ideal.

Reply:
Yes, it is a mixing which is not ideal but it was the best solution for integrity check that I could think of, an example would be something like the following
```
RUN wget https://example.com/myfile.tar.gz
...
RUN gpg --verify myfile.tar.gz.sig myfile.tar.gz
```
These are separate commands which would form separate atomic units, and this function tries to merge the gpg command into a gpg option in the wget atomic unit, without this the smell which checks the download would always flag the wget as not having an integrity check since the check is done with another command

Reply:
Ok I understand. However why can't we do this in the analysis phase? We could improve the detection of the smell "integrity check" to keep track of the files downloaded and see if the files are being verified in another atomic unit. What do you think?

Reply:
Something along the lines if it finds an atomic unit with a checksum command and a file it removes the integrity check from the error list?

Reply:
Yeah something along those lines. I was thinking about going through all the atomic units, and verify the smells normally. However, you would keep two new lists for each unit block: one list with the downloaded files, the other list with the files that were checked (each list was updated in the analysis of the atomic units). In the end of the analysis of each unit block, you would check the files that are both in the downloaded files and checked lists, and remove the integrity check smells related to those files.

Reply:
The removal was made and I will think more clearly on how to handle the smell in the docker_smells branch